### PR TITLE
Speech Recognition for the manual control

### PR DESCRIPTION
**Pull request by ghsdh3409**
### Event Definition
* I define our event handler as followings.
* onSpeechStart: This event is triggered as `Voice.start()`.
* onSpeechEnd: This event is triggered as `Voice.end()`.
* onSpeechPartialResults: This event is triggered as result texts between `onSpeechStart` and `onSpeechEnd`. 

### Problem
* Android triggers events that are onSpeechEnd, onSpeechStart when user start or stop talking.
* After fired onSpeechEnd, speech recognizer does not recognize anymore.
* I fixed this problem to work as above defined.

### CODE REVIEW

File: android/src/main/java/com/wenkesj/voice/VoiceModule.java [hunk 1]
```diff
@@ -30,49 +31,62 @@
   final ReactApplicationContext reactContext;
   private SpeechRecognizer speech = null;
   private boolean isRecognizing = false;
+  private String locale = null;
+  private int numberOfBreakingSentence = 0;
```
Comment (ohtangza) [inline android/src/main/java/com/wenkesj/voice/VoiceModule.java]: Can you elaborate the meaning of this variable naming?

File: android/src/main/java/com/wenkesj/voice/VoiceModule.java [hunk 2]
```diff
@@ -30,49 +31,62 @@
...
-
-        return Locale.getDefault().toString();
-      }
-
       @Override
       public void run() {
         try {
-          speech = SpeechRecognizer.createSpeechRecognizer(self.reactContext);
-        } catch(Exception e) {
+          startListening();
```
Comment (ohtangza) [inline android/src/main/java/com/wenkesj/voice/VoiceModule.java:94]: Seems good!

File: android/src/main/java/com/wenkesj/voice/VoiceModule.java [hunk 3]
```diff
@@ -159,31 +175,52 @@ private void sendEvent(String eventName, @Nullable WritableMap params) {
...
-    WritableMap event = Arguments.createMap();
-    event.putBoolean("error", false);
-    sendEvent("onSpeechEnd", event);
+    if (isRecognizing) {
+      final VoiceModule self = this;
+      Handler mainHandler = new Handler(this.reactContext.getMainLooper());
+      mainHandler.post(new Runnable() {
+        @Override
+        public void run() {
+          speech.cancel();
```
Comment (ohtangza) [inline android/src/main/java/com/wenkesj/voice/VoiceModule.java:218]: So, the end of speech event is captured during the recognition. We re-start the startListening() right?

Comment (ghsdh3409) [reply android/src/main/java/com/wenkesj/voice/VoiceModule.java:218]: Yes. I have seen that after `onEndOfSpeech`, speech recognizer does not recognize anymore. Thus, I restart `startListening()`.

File: android/src/main/java/com/wenkesj/voice/VoiceModule.java (hunk 1 above)
Comment (ghsdh3409) [reply android/src/main/java/com/wenkesj/voice/VoiceModule.java]: Android can break recognizing if a user stops speaking. Our purpose is recognizing before we implicitly stop recognizing. Thus, this variable means that how many times a user stops speaking between start and end of recognizing. Um... `numberOfStopSpeakingInSentence` is more better?

Comment (ohtangza) [reply android/src/main/java/com/wenkesj/voice/VoiceModule.java]: Could you put that explanation in our code as a comment?

File: android/src/main/java/com/wenkesj/voice/VoiceModule.java (hunk 3 above)
Comment (ohtangza) [reply android/src/main/java/com/wenkesj/voice/VoiceModule.java:218]: Got it.

**Review by ohtangza — APPROVED** [review]

**Pull request merged**